beater/auth/authenticator.go

```diff
@@ -0,0 +1,193 @@
+type Authenticator struct {
+	secretToken string
+
+	apikey *apikeyAuth
+}
+
+// Authorizer provides an interface for authorizing an action and resource.
+//
+// TODO(axw) rename interface to Authorizer and method to Authorize.
+type Authorizer interface {
+	// Authorize checks if the client is authorized for the given action and
+	// resource, returning ErrUnauthorized if it is not. Other errors may be
+	// returned, for example because the server cannot communicate with
+	// external systems.
+	Authorize(context.Context, Action, Resource) error
+}
+
+// Resource holds parameters for restricting access that may be checked by
+// Authorizer.Authorize.
+type Resource struct {
+	// AgentName holds the agent name associated with the agent making the
+	// request. This may be empty if the agent is unknown or irrelevant,
+	// such as in a request to the healthcheck endpoint.
+	AgentName string
+
+	// ServiceName holds the service name associated with the agent making
+	// the request. This may be empty if the agent is unknown or irrelevant,
+	// such as in a request to the healthcheck endpoint.
+	ServiceName string
+}
+
+// AuthenticationDetails holds authentication details for a client.
+type AuthenticationDetails struct {
+	// Method holds the authentication kind used.
+	//
+	// Method will be empty for unauthenticated (anonymous) requests when
+	// the server has at least one auth method defined. When the server
+	// has no auth methods defined, this will be MethodNone.
+	Method Method
+
+	// APIKey holds authentication details related to API Key auth.
+	// This will be set when Method is MethodAPIKey.
+	APIKey *APIKeyAuthenticationDetails
+}
+
+// APIKeyAuthenticationDetails holds API Key related authentication details.
+type APIKeyAuthenticationDetails struct {
+	// ID holds the non-secret ID of the API Key.
+	ID string
+
+	// Username holds the username associated with the API Key.
+	Username string
+}
+
+// NewAuthenticator creates an Authenticator with config, authenticating
+// clients with one of the allowed methods.
+func NewAuthenticator(cfg config.AgentAuth) (*Authenticator, error) {
+	b := Authenticator{secretToken: cfg.SecretToken}
+	if cfg.APIKey.Enabled {
+		// do not use username+password for API Key requests
+		cfg.APIKey.ESConfig.Username = ""
+		cfg.APIKey.ESConfig.Password = ""
+		cfg.APIKey.ESConfig.APIKey = ""
+		client, err := elasticsearch.NewClient(cfg.APIKey.ESConfig)
+		if err != nil {
+			return nil, err
+		}
+
+		cache := newPrivilegesCache(cacheTimeoutMinute, cfg.APIKey.LimitPerMin)
+		b.apikey = newApikeyAuth(client, cache)
+	}
+	return &b, nil
+}
+
+// Authenticate authenticates a client given an authentication mtehod and token,
+// returning the authentication details and an Authorizer for authorizing specific
+// actions and resources.
+//
+// Authenticate will return ErrAuthFailed (possibly wrapped) if at least one auth
+// method is configured and the, and no valid credentials have been supplied.
```

Comment:
an errant `and the` from a previous version of the comment?

Reply:
astute :)